### PR TITLE
Bulk insert table

### PR DESCRIPTION
**Pull request by mayhem**
The original motivation for this PR was to get rid of the year-mapping, which could've been done in about 2 lines of code. But I didn't realizes this when I started the project of revamping the year-mapping, so I created a nifty bulk insert table class that manages the creation, population, post-processing, indexing, analyzing and swapping into production of our mapping tables.

The bulk insert table class:
- Can create simple tables
- Can create tables with additional tables that are generated when one SELECT query can generate two output tables.
- Can create destination tables in the musicbrainz DB mapping schema or in the new listenbrainz DB mapping schema.
- Can be chained together so that multiple tables can be created, but they are all swapped into production in one transaction.
- has consistent and useful progress messages, including % complete and how many rows are read/written.

The MBID mapping, mapping releases, canonical recordings and canonical releases are all generated with the top level mbid-mapping command. They are all created in temporary tables and then swapped into production in one transaction, and thus are not longer separate top level commands. The mbid_mapping table now also has a year field for the when that recording was released. (which was the WHOLE POINT of this PR, lol).

UPDATE: Tables have now been renamed:

- mbid_mapping -> canonical_musicbrainz_data
- mbid_mapping_release -> canonical_musicbrainz_data_release
- canonical_recordings -> canonical_recording_redirect
- recording_canonical_release -> canonical_release_redirect

MERGE NOTE: Once approved, this PR should be merged and deployed carefully, as services need to be built and restarted in a particular order.

### CODE REVIEW

File: listenbrainz/mbid_mapping/mapping/bulk_table.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+        log(f"{self.table_name}: post process inserted rows")
+        self._post_process()
+
+        log(f"{self.table_name}: create indexes")
+        self._create_indexes(no_analyze)
+
+        if not no_swap:
+            log(f"{self.table_name}: swap tables and indexes into production.")
+            self._swap_into_production()
```
Comment (amCap1712) [inline listenbrainz/mbid_mapping/mapping/bulk_table.py:385]: remove `_`

File: listenbrainz/mbid_mapping/mapping/canonical_musicbrainz_data_release.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    """
+        This class creates the MBID mapping release table.
+
+        For documentation on what each of the functions in this class does, please refer
+        to the BulkInsertTable docs.
+    """
+
+    def __init__(self, mb_conn, lb_conn=None, batch_size=None):
+        super().__init__("mapping.canonical_musicbrainz_data_release", mb_conn, lb_conn, batch_size)
+        self.release_index = {}
```
Comment (amCap1712) [inline listenbrainz/mbid_mapping/mapping/canonical_musicbrainz_data_release.py:24]: Not necessary but could use a `set` here.

File: listenbrainz/mbid_mapping/mapping/canonical_release_redirect.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    def get_post_process_queries(self):
+        return ["""INSERT INTO mapping.canonical_release_redirect_tmp (recording_mbid, release_mbid)
+                        SELECT recording_mbid
+                             , canonical_release_mbid AS release_mbid
+                          FROM mapping.canonical_recording_redirect_tmp""",
+                """INSERT INTO mapping.canonical_release_redirect_tmp (recording_mbid, release_mbid)
+                                 SELECT recording_mbid
+                                      , release_mbid
+                                   FROM mapping.canonical_musicbrainz_data_tmp""",
+                """COMMIT"""]
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/mbid_mapping/mapping/canonical_release_redirect.py:32]: The normal insert process does, but these are post-process steps, for which there are no commits. Without this, rotating large indexes causes issues.

File: listenbrainz/mbid_mapping/mapping/bulk_table.py (hunk 1 above)
Comment (mayhem) [reply listenbrainz/mbid_mapping/mapping/bulk_table.py:385]: Ah yeah.

**Review by amCap1712 — APPROVED** [review]
LGTM, Thanks!

File: listenbrainz/mbid_mapping/mapping/bulk_table.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+            Run queries that do post insert cleanup/processing. These queries are run
+            BEFORE indexes are created. Can be used for deduping created data, for instance.
+        """
+
+        conn = self.lb_conn if self.lb_conn is not None else self.mb_conn
+        try:
+            with conn.cursor() as curs:
+                for query in self.get_post_process_queries():
+                    curs.execute(query)
+                conn.commit()
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/mbid_mapping/mapping/bulk_table.py:191]: ??

**Pull request merged**